DMGC-IPS-01/DMGC-IPS-01_ATTINY85.ino

```diff
@@ -164,25 +191,71 @@ void loop() {
         }
         delay(DEBOUNCE_DELAY);
       }
-      
+
       if(brightness_change_flag==0){
         if(color_type==COLOR_QTY-1){
           color_type=0;
+          rainbow_current=RAINBOW_START;
         }else{
           color_type = color_type+1;
         }
       }
       // Save brightness and color setting to EPROM
-      EEPROM.put(0,brightness);
-      EEPROM.put(1,color_type);
+      EEPROM.update(0,brightness);
+      EEPROM.update(1,color_type);
       brightness_change_flag = 0;
       }
     }
+
+    // Update LED outputs
+    outputLED(color_type);
+
+    // Sleep if showing a static color or static rainbow
+    if (color_type <= RAINBOW_START_IDX) {
+      sleep(true);
+    } else {
+      // Delay just a bit so we're not refreshing the LEDs so fast
+      delay(LOOP_DELAY);
+    }
   }
 
 void outputLED(uint8_t x){
-  for(int i=0; i<NUMPIXELS; i++) { // For each pixel...
-    pixels.setPixelColor(i, pixels.Color(red[x],green[x],blue[x]));
-    pixels.show();   // Send the updated pixel colors to the hardware.
-  }  
+  // Rainbow logic; similar to NeoPixel library's, except with first two pixels swapped
+  if (x >= RAINBOW_START_IDX) {
+    for (uint16_t i=0; i<NUMPIXELS; i++) { // For each pixel...
+      uint16_t j = i; // Swap left and up
+      if (j == u) {
+        j = l;
+      } else if (j == l) {
+        j = u;
+      }
+      uint16_t hue = rainbow_current + (i * 65536) / NUMPIXELS;
+      uint32_t color = pixels.ColorHSV(hue, RAINBOW_SAT, RAINBOW_VAL);
+      pixels.setPixelColor(j, pixels.gamma32(color));
+    }
+    rainbow_current += rainbow[x];
+  } else {
+    for(int i=0; i<NUMPIXELS; i++) { // For each pixel..
+      pixels.setPixelColor(i, pixels.Color(red[x],green[x],blue[x]));
+    }
+  }
+  pixels.show();   // Send the updated pixel colors to the hardware.
+}
+
+// ISR - called when push is pushed during sleep to wake up
+ISR(PCINT0_vect) {
+  GIMSK &= ~0b00100000;  // Turn off pin change interrupts
+  sleep_disable();
+}
+
+// Sleeps to save ~25 mW
+void sleep(bool can_wake) {
+  sleep_enable();
+  if (can_wake) { // Enable interrupts so we can wake up
+    noInterrupts();
+    GIMSK |= 0b00100000;  // Turn on pin change interrupts
+    PCMSK |= 1 << pushbtn; // For pin 3 - pushbtn
+    interrupts();
+  }
+  sleep_cpu();
 }

```

Comment:
Sleep stuff. Referenced from a random sketch I found on the internet: https://github.com/urish/tiny-simon/blob/master/tiny-simon.ino